tests/test_feedexport.py

```diff
@@ -1587,16 +1590,60 @@ def test_export_items(self):
         header = self.MyItem.fields.keys()
         yield self.assertExported(items, header, rows, settings=Settings(settings))
 
+    @defer.inlineCallbacks
+    def test_export_file_size_trigger(self):
+        items = [
+            self.MyItem({'foo': 'bar1', 'egg': 'spam1', 'baz': 'quux1'}),
+            self.MyItem({'foo': 'bar2', 'egg': 'spam2', 'baz': 'quux2'}),
+        ]
+        fmt_to_limits = {
+            'json': ('40B', '90B'),
+            'jsonlines': ('40B', '90B'),
+            'xml': ('100B', '160B'),
+            'csv': ('30B', '40B'),
+            'marshal': ('30B', '70B'),
+            'pickle': ('50B', '110B'),
+        }
+        for fmt, limits in fmt_to_limits.items():
+            #  +1 for empty files after batch has been triggered
+            for limit, expected in {'0B': 1, '1B': 2 + 1, limits[0]: 2 + 1, limits[1]: 1 + 1}.items():
+                settings = {
+                    'FEEDS': {
+                        os.path.join(self._random_temp_filename(), fmt, self._file_mark): {'format': fmt},
+                    },
+                    'FEED_EXPORT_BATCH_FILE_SIZE': limit,
+                }
+                data = yield self.exported_data(items, settings)
+                self.assertEqual(expected, len(data[fmt]))
+
+    @defer.inlineCallbacks
+    def test_export_duration_trigger(self):
+        items = [self.MyItem({'foo': 'bar1', 'egg': 'spam1', 'baz': 'quux1'})]
+        for fmt in ('csv', 'json', 'jsonlines', 'marshal', 'pickle'):
+            #  +1 for empty files after batch has been triggered
+            for delay, expected in {'0:0:0': 1, '0:0:0.01': 2 + 1, '0:0:0.1': 2 + 1, '0:0:1.2': 1}.items():
+                settings = {
+                    'FEEDS': {
+                        os.path.join(self._random_temp_filename(), fmt, self._file_mark): {'format': fmt},
+                    },
+                    'FEED_EXPORT_BATCH_DURATION': delay,
+                    'DOWNLOAD_DELAY': 0.5,
+                }
+                data = yield self.exported_data(items, settings, multiple_urls=2)
+                self.assertEqual(expected, len(data[fmt]))
```

Comment:
This test, with a single item, seems to indirectly demonstrate that the current implementation will cause empty files to be created sometimes.

I think this is undesired behavior for the ideal batch implementation, i.e. ideally we should only create files with items. But I also think this behavior is OK for an initial implementation, and I think it is a consequence of pre-existing batch-handling code.

Reply:
Yes I think in the original batch trigger PR this was pointed out. Here: [https://github.com/scrapy/scrapy/pull/4434#issuecomment-658672328](https://github.com/scrapy/scrapy/pull/4434#issuecomment-658672328).

Reply:
This is slightly different, though. The initial implementation creates 1 extra empty file. Somehow this test, with 1 item, can create 3 different files. I’m actually not sure how this happens, but it seems to be happening because the test passes. Any idea why?

Reply:
Oh I see, I should have put a comment explaining it. So to simulate item scraping delays I am using `DOWNLOAD_DELAY` to delay url downloads. So as to get one item per download. That's why I am using `multiple_urls` argument in the `exported_data` function. You suggested that trick for testing hard limit duration :)

Reply:
So I am actually getting total of 2 items, one from each url download.

Reply:
:facepalm:

That makes total sense, sorry for the noise!